include/xsparse/level_capabilities/co_iteration.hpp

```diff
@@ -0,0 +1,200 @@
+#ifndef XSPARSE_CO_ITERATION_HPP
+#define XSPARSE_CO_ITERATION_HPP
+
+#include <vector>
+#include <tuple>
+#include <algorithm>
+#include <stdexcept>
+
+namespace xsparse::level_capabilities
+{
+    template <class F, class IK, class PK, class Levels, class Is>
+    class Coiterate;
+
+    template <class F, class IK, class PK, class... Levels, class... Is>
+    class Coiterate<F, IK, PK, std::tuple<Levels...>, std::tuple<Is...>>
+    {
+    private:
+        std::tuple<Levels&...> const m_levelsTuple;
+        F const m_comparisonHelper;
+
+    public:
+        explicit inline Coiterate(F f, Levels&... levels)
+            : m_levelsTuple(std::tie(levels...))
+            , m_comparisonHelper(f)
+        {
+            if (![](auto const& first, auto const&... rest)
+                { return ((first == rest) && ...); }(levels.size()...))
+            {
+                throw std::runtime_error("level sizes should be same");
```

Comment:
I'd replace with `std::invalid_argument`.

Reply:
Yeah, I'll do that.